### PR TITLE
Initial typescript declarations and tests

### LINKED ISSUE #142: Typescript definitions
Hello @taion,

Are you interested in [bundling a typescript definition file in the package](http://www.typescriptlang.org/docs/handbook/declaration-files/publishing.html)?  I have created/documented one and could submit it with a PR if you like.

### PR DESCRIPTION
**Pull request by rosskevin**
Closes #142 

I have been using these definitions locally - so far so good.

Many of the types have been documented and are solid based on my use, though some of the others have been less traveled and are typed as `any`.  While `any` is relevant in some situations, often it is bailing out.  So these can be improved as we go.

In `tests/typescript` I have added the sample usage from the readme, as well as added the `typescript` script in for testing.  This `test-usage.tsx` file can be expanded, or other files added to the directory as holes are found.  

I think this setup should be fairly easy to keep up/validate.  I expect others will find holes and PR small changes from here as well.

### CODE REVIEW

**Comment by taion** [comment]
I’ve been on the road. Will look next week.

File: src/index.d.ts [hunk 1]
```diff
@@ -0,0 +1,350 @@
...
+   *  - A nully value to call the next transition hook and use its return value, if present, or
+   *    else to allow the transition
+   *  - A promise that resolves to any of the above values, to allow or block the transition
+   *    once the promise resolves
+   *
+   * @see https://github.com/4Catalyzer/farce#transition-hooks
+   */
+  type TransitionHook = (
+    location: Location,
+  ) => undefined | (boolean | string | Promise<boolean | string>)
```
Comment (itajaja) [inline src/index.d.ts]: is the parens needed here?

File: src/index.d.ts [hunk 2]
```diff
@@ -0,0 +1,350 @@
...
+
+  interface CreateBrowserRouterArgs {
+    render?: (args: RenderArgs) => React.ReactElement<any>
+    renderPending?: (args: RenderArgs) => React.ReactElement<any>
+    renderReady?: (args: RenderReadyArgs) => React.ReactElement<any>
+    renderError?: (args: RenderArgs) => React.ReactElement<any>
+    [key: string]: any
+  }
+
+  function createBrowserRouter({
```
Comment (itajaja) [inline src/index.d.ts]: does the spread/destructuring functionally do anything here? wouldn't be the same to do `options: CreateBrowserRouterArgs`?

File: src/index.d.ts (hunk 1 above)
Comment (rosskevin) [reply src/index.d.ts]: It's a matter of style, my local prettier formatted this so it is consistent

File: src/index.d.ts (hunk 2 above)
Comment (rosskevin) [reply src/index.d.ts]: Not needed, it was when I copied in the signature. I'll remove

File: src/index.d.ts [hunk 3]
```diff
@@ -0,0 +1,344 @@
...
+    renderError?: (args: RenderArgs) => React.ReactElement<any>
+    [key: string]: any
+  }
+
+  function createBrowserRouter(options: CreateBrowserRouterArgs): BrowserRouter
+
+  // Improve these `any`s as needed
+  type FarceRouter = any
+
+  function createFarceRouter({
```
Comment (itajaja) [inline src/index.d.ts]: another one here :) why not making a type out of these args and use that instead of `any`?

Comment (rosskevin) [reply src/index.d.ts]: As noted, this is a starting point.  Feel free to enhance.  I have typed everything I use (and more), now it's time for everyone to make it better.

**Comment by taion** [comment]
@itajaja Do the definitions look good to you?

**Comment by itajaja** [comment]
definitions are 👍  The only thing is that the `ts` test files are actually not used during CI, so there is a risk they'd go out of sync. The rest looks good!

**Comment by itajaja** [comment]
maybe adding the `typescript` command during test would at lest validate the test files statically?

**Comment by rosskevin** [comment]
@itajaja `test` script is already modified to execute `typescript`.

**Comment by taion** [comment]
@rosskevin I'm sorry for the delay here. We've been really busy lately. I'm going to look to try to merge this next week.

File: src/index.d.ts [hunk 4]
```diff
@@ -0,0 +1,344 @@
...
+     */
+    key?: string
+    /**
+     * the path name; as on window.location e.g. '/foo'
+     */
+    pathname: string
+    /**
+     * map version of search string
+     */
+    query: Map
```
Comment (taion) [inline src/index.d.ts]: this is a little stricter – it's string -> string

File: src/index.d.ts [hunk 5]
```diff
@@ -0,0 +1,344 @@
...
+  }
+
+  /**
+   * The shape might be different with a custom matcher or history enhancer, but the default matcher
+   * assumes and provides this shape. As such, this validator is purely for user convenience and
+   * should not be used internally.
+   */
+  interface Match {
+    location: Location
+    params: Map
```
Comment (taion) [inline src/index.d.ts]: this is a little stricter – it's string -> string (i think?)

File: src/index.d.ts [hunk 6]
```diff
@@ -0,0 +1,344 @@
...
+   */
+  interface Matcher {
+    match: Function
+    getRoutes: Function
+    isActive: Function
+    /**
+     * Returns the path string for a pattern of the same format as a route path and a object of the
+     * corresponding path parameters
+     */
+    format: (pattern: any, params: Map) => any
```
Comment (taion) [inline src/index.d.ts]: This will support non-string things in params though

File: src/index.d.ts [hunk 7]
```diff
@@ -0,0 +1,344 @@
...
+  }
+
+  /**
+   * Lenient arg version of location using in #push and #replace.
+   */
+  type LocationArg = Pick<Location, 'pathname'> &
```
Comment (taion) [inline src/index.d.ts]: the union of this and "string" below is referred to as a "location descriptor" in the other docs: https://github.com/4Catalyzer/farce#locations-and-location-descriptors

File: src/index.d.ts [hunk 8]
```diff
@@ -0,0 +1,344 @@
...
+  type TransitionHook = (
+    location: Location,
+  ) => undefined | (boolean | string | Promise<boolean | string>)
+
+  class Router {
+    /**
+     * Navigates to a new location
+     * @see farce
+     */
+    public push: (location: string | LocationArg) => void
```
Comment (taion) [inline src/index.d.ts]: can you define the arg here as a `LocationDescriptor`?

Comment (taion) [inline src/index.d.ts]: this actually does return whatever the action dispatch returns, but i'm guessing you wouldn't want to use that return value

File: src/index.d.ts [hunk 9]
```diff
@@ -0,0 +1,344 @@
...
+     * @see farce
+     */
+    public replace: (location: string | LocationArg) => void
+    /**
+     * Moves delta steps in the history stack
+     * @see farce
+     */
+    public go: (delta: number) => void
+
+    public createHref: Function
```
Comment (taion) [inline src/index.d.ts]: `(location: LocationDescriptor) => string`

File: src/index.d.ts [hunk 10]
```diff
@@ -0,0 +1,344 @@
...
+    public createLocation: Function
```
Comment (taion) [inline src/index.d.ts]: `(location: LocationDescriptor) => Location`

File: src/index.d.ts [hunk 11]
```diff
@@ -0,0 +1,344 @@
+// Type definitions for found v0.3.5
+// Project: https://github.com/4Catalyzer/found
+// Definitions by: Kevin Ross <https://github.com/rosskevin/>
+declare module 'found' {
+  import * as React from 'react'
+
+  interface Map {
```
Comment (taion) [inline src/index.d.ts]: is it standard to call this `Map`? could this be ambiguous w/ES6 `Map`?

**Comment by rosskevin** [comment]
Thanks for the review @taion - all look like good/appropriate changes.  I'm underwater with some other efforts, I suggest merging and allowing other contributors to make it better.  I'm pretty buried so feel free to do as you like, it's working for me for the last month so I don't have the impetus to pick it back up.

**Comment by taion** [comment]
Yeah, same here, sorry. We can pick this up and try to get this merged soon.

**Comment by TomMarius** [comment]
Any news on this?

**Comment by khrystalz** [comment]
@rosskevin Great work!
Since there hasn't been any update on this for a while: have you thought about contributing your definitions to the [DefinitelyTyped](https://github.com/DefinitelyTyped/DefinitelyTyped) project?

**Comment by taion** [comment]
Gah I totally forgot about this.

**Comment by taion** [comment]
I updated this PR to be mergeable again (including https://github.com/4Catalyzer/found/pull/174) and I addressed my own PR comments.

Can someone here take a look? If I can get a quick sign-off here, I'd be happy to merge and cut a release.

**Pull request merged**